src/com/ms/silverking/cloud/dht/daemon/storage/FileSegment.java

```diff
@@ -23,243 +23,255 @@
     }
-    
-    public static FileSegment openReadOnly(File nsDir, int segmentNumber, int dataSegmentSize, 
-                                           NamespaceOptions nsOptions, SegmentIndexLocation segmentIndexLocation, SegmentPrereadMode segmentPrereadMode) throws IOException {
-        return open(nsDir, segmentNumber, AccessMode.ReadOnly, dataSegmentSize, SyncMode.NoSync, nsOptions, segmentIndexLocation, segmentPrereadMode);
+
+    switch (accessMode) {
+    case ReadOnly:      // fall through
+    case ReadIndexOnly:
+      dataMapMode = MapMode.READ_ONLY;
+      fileOpenMode = roFileMode;
+      break;
+    case Update:
+      dataMapMode = MapMode.READ_WRITE;
+      fileOpenMode = syncModeToFileOpenMode(syncMode);
+      break;
+    default:
+      throw new RuntimeException("Unexpected mode: " + accessMode);
     }
-    
-    private static FileSegment open(File nsDir, int segmentNumber, AccessMode accessMode, int dataSegmentSize, 
-                                    SyncMode syncMode, NamespaceOptions nsOptions, 
-                                    SegmentIndexLocation segmentIndexLocation, SegmentPrereadMode segmentPrereadMode) 
-                                            throws IOException {
-        RandomAccessFile    raFile;
-        ByteBuffer          dataBuf;
-        ByteBuffer          rawHTBuf;
-        String              fileOpenMode;
-        MapMode             dataMapMode;
-        FileSegmentMetaData         fsm;
-        InvalidatedOffsetsElement   ioe;
-        KeyToOffsetMapElement       ktome;
-        OffsetListsElement          ole;
-        
-        //Log.warningf("open %s %d", nsDir.toString(), segmentNumber);
-        
-        if (accessMode == AccessMode.ReadIndexOnly && nsOptions.getVersionMode() == NamespaceVersionMode.SINGLE_VERSION) {
-            accessMode = AccessMode.ReadOnly;
-            // Presently, the index does not contain sufficient information to allow this in SINGLE_VERSION mode
-        }
+    raFile = new RandomAccessFile(fileForSegment(nsDir, segmentNumber), fileOpenMode);
+    if (false) {
+      // This would allow us to force each segment onto the heap
+      // FUTURE: consider whether we want to keep this option
+      byte[] _bufArray;
 
-        switch (accessMode) {
-        case ReadOnly:      // fall through
-        case ReadIndexOnly:
-            dataMapMode = MapMode.READ_ONLY;
-            fileOpenMode = roFileMode;
-            break;
-        case Update:
-            dataMapMode = MapMode.READ_WRITE;
-            fileOpenMode = syncModeToFileOpenMode(syncMode);
-            break;
-        default: throw new RuntimeException("Unexpected mode: "+ accessMode);
+      _bufArray = new byte[dataSegmentSize];
+      raFile.read(_bufArray);
+      dataBuf = ByteBuffer.wrap(_bufArray);
+    } else {
+      if (accessMode != AccessMode.ReadIndexOnly) {
+        dataBuf = raFile.getChannel().map(dataMapMode, 0, dataSegmentSize);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *getChannel*. Currently, there are execution paths that do not contain closure statements, for example, when FileChannel.map throws an exception. Either a) close the object returned by *getChannel()* in a try-finally block or b) close the resource by declaring the object returned by *getChannel()* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).